datasource/src/main/scala/quasar/physical/mongo/decoder.scala

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2014–2018 SlamData Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.physical.mongo
+
+import slamdata.Predef._
+
+import eu.timepit.refined.auto._
+import java.time._
+import org.bson._
+import qdata._
+import qdata.QType._
+import scala.collection.JavaConverters._
+import spire.math.Real
+
+object decoder {
+  val qdataDecoder: QDataDecode[BsonValue] = new QDataDecode[BsonValue] {
+    override def tpe(bson: BsonValue): QType = bson.getBsonType() match {
+      case BsonType.DOCUMENT => QObject
+      case BsonType.ARRAY => QArray
+      case BsonType.STRING => QString
+      case BsonType.INT32 => QLong
+      case BsonType.INT64 => QLong
+      case BsonType.DOUBLE => QDouble
+      case BsonType.DECIMAL128 => QReal
+      case BsonType.BOOLEAN => QBoolean
+      case BsonType.OBJECT_ID => QString
+      case BsonType.DB_POINTER => QString
+      case BsonType.BINARY => QString
+      case BsonType.DATE_TIME => QOffsetDateTime
+      case BsonType.SYMBOL => QString
+      case BsonType.REGULAR_EXPRESSION => QString
+      case BsonType.JAVASCRIPT => QString
+      case BsonType.NULL => QNull
+      case BsonType.TIMESTAMP => QOffsetDateTime
+      case BsonType.UNDEFINED => QNull
+      case BsonType.JAVASCRIPT_WITH_SCOPE => QString
+      case BsonType.MAX_KEY => QString
+      case BsonType.MIN_KEY => QString
+      case BsonType.END_OF_DOCUMENT => QNull
+    }
+
+    type ArrayCursor = List[BsonValue]
+
+    override def getArrayCursor(bson: BsonValue): ArrayCursor = bson match {
+      case arr: BsonArray => {
+        arr.getValues().asScala.toList
+      }
+    }
+    override def getArrayAt(cursor: ArrayCursor): BsonValue = cursor match {
+      case (hd :: _) => hd
```

Comment:
usually we use it without brackets: `case hd :: _ => hd` (below there are some similar usages)